### PR TITLE
Add Uri StringSyntaxAttribute syntaxes

### PR DESCRIPTION
**Pull request by blouflashdb**
# Add Uri StringSyntaxAttribute syntaxes

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

Adds uri StringSyntaxAttribute syntaxes to all public APIs where an uri is expected. 

Closes Uri Task in #44535

### CODE REVIEW

**Comment by blouflashdb** [comment]
Well somehow I can't use it in HubConnectionBuilderHttpExtensions.cs for some reason.
Check failure on line 26 in src/SignalR/clients/csharp/Client/src/HubConnectionBuilderHttpExtensions.cs

Azure Pipelines
/ aspnetcore-ci (Build Build: Linux ARM)
src/SignalR/clients/csharp/Client/src/HubConnectionBuilderHttpExtensions.cs#L26
src/SignalR/clients/csharp/Client/src/HubConnectionBuilderHttpExtensions.cs(26,131): error CS0122: (NETCORE_ENGINEERING_TELEMETRY=Build) 'StringSyntaxAttribute' is inaccessible due to its protection level

**Comment by JamesNK** [comment]
The SignalR client also targets .NET Framework, which doesn't have that attribute.

To work around this, you can copy the attribute's source code from [StringSyntaxAttribute.cs](https://github.com/dotnet/runtime/blob/8ed6b13a103de3a1b529dea23fb0e2dfac94fc1b/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs) and an internal polyfill type in the Shared/CodeAnalysis directory - https://github.com/dotnet/aspnetcore/tree/1b53b22a866eb4bc58f2d5552db7f3ac08f1a952/src/Shared/CodeAnalysis. Then add a link to the shared source file from the SignalR client project.

**Comment by blouflashdb** [comment]
Thanks, I think I got it.

File: src/Components/Components/src/NavigationManager.cs [hunk 1]
```diff
@@ -103,7 +104,7 @@ protected set
     /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
     /// (as returned by <see cref="BaseUri"/>).</param>
     /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
-    public void NavigateTo(string uri, bool forceLoad) // This overload is for binary back-compat with < 6.0
+    public void NavigateTo([StringSyntax(StringSyntaxAttribute.Uri, UriKind.RelativeOrAbsolute)] string uri, bool forceLoad) // This overload is for binary back-compat with < 6.0
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:107]: None of the StringSyntax usages in dotnet/runtime - https://cs.github.com/dotnet/runtime?q=StringSyntaxAttribute.Uri - are passing UriKind itself as an argument. For example, https://github.com/dotnet/runtime/blob/960e4d723c27a5407dc691d06e546bc455a9c4a5/src/libraries/System.Private.Uri/src/System/Uri.cs#L408

Also, passing in `UriKind.RelativeOrAbsolute` doesn't seem useful as I think that would be the default.

@stephentoub @CyrusNajmabadi What should the argument be with `StringSyntaxAttribute.Uri`? Could whatever future tooling that uses that parameter support both a string that refers to another parameter or `UriKind` value?

Comment (JamesNK) [reply src/Components/Components/src/NavigationManager.cs:107]: [`StringSyntaxAttribute.Regex` in dotnet/runtime](https://cs.github.com/dotnet/runtime?q=StringSyntaxAttribute.Regex) is happy with either which answers my question!

Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:107]:
```suggestion
    public void NavigateTo([StringSyntax(StringSyntaxAttribute.Uri)] string uri, bool forceLoad) // This overload is for binary back-compat with < 6.0
```

File: src/Components/Components/src/NavigationManager.cs [hunk 2]
```diff
@@ -113,7 +114,7 @@ protected set
     /// (as returned by <see cref="BaseUri"/>).</param>
     /// <param name="forceLoad">If true, bypasses client-side routing and forces the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
     /// <param name="replace">If true, replaces the current entry in the history stack. If false, appends the new entry to the history stack.</param>
-    public void NavigateTo(string uri, bool forceLoad = false, bool replace = false)
+    public void NavigateTo([StringSyntax(StringSyntaxAttribute.Uri, UriKind.RelativeOrAbsolute)] string uri, bool forceLoad = false, bool replace = false)
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:117]:
```suggestion
    public void NavigateTo([StringSyntax(StringSyntaxAttribute.Uri)] string uri, bool forceLoad = false, bool replace = false)
```

File: src/Components/Components/src/NavigationManager.cs [hunk 3]
```diff
@@ -139,7 +140,7 @@ public void NavigateTo(string uri, bool forceLoad = false, bool replace = false)
     /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
     /// (as returned by <see cref="BaseUri"/>).</param>
     /// <param name="options">Provides additional <see cref="NavigationOptions"/>.</param>
-    public void NavigateTo(string uri, NavigationOptions options)
+    public void NavigateTo([StringSyntax(StringSyntaxAttribute.Uri, UriKind.RelativeOrAbsolute)] string uri, NavigationOptions options)
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:143]:
```suggestion
    public void NavigateTo([StringSyntax(StringSyntaxAttribute.Uri)] string uri, NavigationOptions options)
```

File: src/Components/Components/src/NavigationManager.cs [hunk 4]
```diff
@@ -155,7 +156,7 @@ public void NavigateTo(string uri, NavigationOptions options)
     // already override this, so the framework needs to keep using it for the cases when only pre-6.0 options are used.
     // However, for anyone implementing a new NavigationManager post-6.0, we don't want them to have to override this
     // overload any more, so there's now a default implementation that calls the updated overload.
-    protected virtual void NavigateToCore(string uri, bool forceLoad)
+    protected virtual void NavigateToCore([StringSyntax(StringSyntaxAttribute.Uri, UriKind.RelativeOrAbsolute)] string uri, bool forceLoad)
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:159]:
```suggestion
    protected virtual void NavigateToCore([StringSyntax(StringSyntaxAttribute.Uri)] string uri, bool forceLoad)
```

File: src/Components/Components/src/NavigationManager.cs [hunk 5]
```diff
@@ -164,7 +165,7 @@ protected virtual void NavigateToCore(string uri, bool forceLoad)
     /// <param name="uri">The destination URI. This can be absolute, or relative to the base URI
     /// (as returned by <see cref="BaseUri"/>).</param>
     /// <param name="options">Provides additional <see cref="NavigationOptions"/>.</param>
-    protected virtual void NavigateToCore(string uri, NavigationOptions options) =>
+    protected virtual void NavigateToCore([StringSyntax(StringSyntaxAttribute.Uri, UriKind.RelativeOrAbsolute)] string uri, NavigationOptions options) =>
```
Comment (JamesNK) [inline src/Components/Components/src/NavigationManager.cs:168]:
```suggestion
    protected virtual void NavigateToCore([StringSyntax(StringSyntaxAttribute.Uri)] string uri, NavigationOptions options) =>
```

File: src/Components/WebView/WebView/src/WebViewManager.cs [hunk 6]
```diff
@@ -62,7 +63,7 @@ public WebViewManager(IServiceProvider provider, Dispatcher dispatcher, Uri appB
     /// client-side routing.
     /// </summary>
     /// <param name="url">The URL, which may be absolute or relative to the application root.</param>
-    public void Navigate(string url)
+    public void Navigate([StringSyntax(StringSyntaxAttribute.Uri, UriKind.RelativeOrAbsolute)] string url)
```
Comment (JamesNK) [inline src/Components/WebView/WebView/src/WebViewManager.cs:66]:
```suggestion
    public void Navigate([StringSyntax(StringSyntaxAttribute.Uri)] string url)
```

**Review by JamesNK — COMMENTED** [review]
RelativeOrAbsolute would be the default assumption so there is no need to explicitly specify it.

Otherwise, LGTM

**Review by JamesNK — APPROVED** [review]

**Pull request merged**